pykeops/torch/nystrom/nystrom.py

```diff
@@ -0,0 +1,265 @@
+
+        K_ij = K_ij @ torch.diag(torch.ones(K_ij.shape[1]))  # make 1 on diag only
+
+        K_ij.backend = self.backend
+        return K_ij
+
+    def _Gauss_block_sparse_pre(
+        self, x: torch.tensor, y: torch.tensor, K_ij: LazyTensor
+    ):
+        """
+        Helper function to preprocess data for block-sparse reduction
+        of the Gaussian kernel
+
+        Args:
+            x[np.array], y[np.array] = arrays giving rise to Gaussian kernel K(x,y)
+            K_ij[LazyTensor_n] = symbolic representation of K(x,y)
+            eps[float] = size for square bins
+        Returns:
+            K_ij[LazyTensor_n] = symbolic representation of K(x,y) with
+                                set sparse ranges
+        """
+        # labels for low dimensions
+
+        if x.shape[1] < 4 or y.shape[1] < 4:
+
+            x_labels = grid_cluster(x, self.eps)
+            y_labels = grid_cluster(y, self.eps)
+            # range and centroid per class
+            x_ranges, x_centroids, _ = cluster_ranges_centroids(x, x_labels)
+            y_ranges, y_centroids, _ = cluster_ranges_centroids(y, y_labels)
+        else:
+            # labels for higher dimensions
+
+            x_labels, x_centroids = self._KMeans(x)
+            y_labels, y_centroids = self._KMeans(y)
+            # compute ranges
+            x_ranges = cluster_ranges(x_labels)
+            y_ranges = cluster_ranges(y_labels)
+
+        # sort points
+        x, x_labels = sort_clusters(x, x_labels)
+        y, y_labels = sort_clusters(y, y_labels)
+        # Compute a coarse Boolean mask:
+        D = torch.sum((x_centroids[:, None, :] - y_centroids[None, :, :]) ** 2, 2)
+        keep = D < (self.mask_radius) ** 2
+        # mask -> set of integer tensors
+        ranges_ij = from_matrix(x_ranges, y_ranges, keep)
+        K_ij.ranges = ranges_ij  # block-sparsity pattern
+
+        return K_ij
+
+    def _KMeans(self, x: torch.tensor):
+        """KMeans with Pykeops to do binning of original data.
+        Args:
+            x[np.array] = data
+            k_means[int] = number of bins to build
+            n_iter[int] = number iterations of KMeans loop
+        Returns:
+            labels[np.array] = class labels for each point in x
+            clusters[np.array] = coordinates for each centroid
+        """
+
+        N, D = x.shape
+        clusters = torch.clone(x[: self.k_means, :])  # initialization of clusters
+        x_i = LazyTensor(x[:, None, :])
+
+        for i in range(self.n_iter):
+
+            clusters_j = LazyTensor(clusters[None, :, :])
+            D_ij = ((x_i - clusters_j) ** 2).sum(-1)  # points-clusters kernel
+            labels = D_ij.argmin(axis=1).reshape(N)  # Points -> Nearest cluster
+            Ncl = torch.bincount(labels)  # Class weights
+            for d in range(D):  # Compute the cluster centroids with np.bincount:
+                clusters[:, d] = torch.bincount(labels, weights=x[:, d]) / Ncl
+
+        return labels, clusters
```

Comment:
I assume that you will factor this with the numpytools/torchtools implementation?

Reply:
done